### PR TITLE
Require prism >= 0.19.0

### PR DESCRIPTION
**Pull request by tompng**
Remove StringNode workaround
Simplify CallNode attribute assign check (`CallNode#attribute_write?` is added)
Fix for added nodes `NumberedParametersNode` `CallTargetNode` `IndexTargetNode` `ImplicitRestNode`

### CODE REVIEW

File: lib/repl_type_completor.rb [hunk 1]
```diff
@@ -123,13 +126,15 @@ def analyze_code(code, binding = Object::TOPLEVEL_BINDING)
     end
 
     def find_target(node, position)
+      # Skip because NumberedParametersNode#location gives location of whole block
+      return if node.type == :numbered_parameters_node
```
Comment (st0012) [inline lib/repl_type_completor.rb:130]: Can this be a `when` in the case statement below?

**Review by st0012 — COMMENTED** [review]
If I understand it correctly, after this change we'll be supporting both Prism `0.18` and `0.19`? If that's the case, maybe we should just move towards `0.19` as people should always upgrade Prism, especially before Ruby 3.3?
In the meantime, we can point Prism to GH source in the Gemfile.

**Comment by tompng** [comment]
I was thinking of:
1. Support both(prism head and 0.18.0) for now
2. Switch to 0.19.0, remove code for 0.18.0
3. Release repl_type_completor

Mainly to make CI green. (but it makes code complicated...)

Using github source in Gemfile seems good, I'll try it.

File: lib/repl_type_completor.rb [hunk 2]
```diff
@@ -68,17 +68,18 @@ def analyze_code(code, binding = Object::TOPLEVEL_BINDING)
...
         return unless args_node.is_a?(Prism::ArgumentsNode) && args_node.arguments.size == 1
 
-        content = code.byteslice(target_node.opening_loc.end_offset..)
         case call_node.name
         when :require
-          [:require, content]
+          [:require, target_node.content]
         when :require_relative
-          [:require_relative, content]
+          [:require_relative, target_node.content]
```
Comment (st0012) [inline lib/repl_type_completor.rb:82]: I think these can be

```rb
case call_node.name
when :require, :require_relative
  [call_node.name, target_node.content]
when
  # ....
```

File: lib/repl_type_completor/type_analyzer.rb [hunk 3]
```diff
@@ -925,8 +923,7 @@ def assign_parameters(node, scope, args, kwargs)
       node.posts.zip posts do |n, v|
         assign_required_parameter n, v, scope
       end
-      if node.rest&.name
-        # node.rest is Prism::RestParameterNode
+      if node.rest.is_a?(Prism::RestParameterNode) && node.rest.name
```
Comment (st0012) [inline lib/repl_type_completor/type_analyzer.rb:926]: There's a `if node.rest` condition a few lines above, which seems to already expect it to be `Prism::RestParameterNode`? Perhaps we can merge that and this together?

Comment (tompng) [reply lib/repl_type_completor/type_analyzer.rb:926]: I removed the confusing comment in the `if node.rest` condition and add a comment about `Prism::ImplicitRestNode` that does not have name.

The `if node.rest` above part is separating given args to reqs, posts, rest
`method do |a,b=1,*c| ` and given args `[Integer, String, Symbol]` to  `reqs = [Integer], opts = [String], posts = [], rest = [Symbol]`.

This part is assigning each of them.
```ruby
node.requireds.zip{assign}
node.optionals.zip{assign}
node.posts.zip{assign}
if node.rest; assign; end
```
I want to separate them, not to merge it together.

Comment (st0012) [reply lib/repl_type_completor/type_analyzer.rb:926]: I see, thanks for the clear explanation 👍

**Review by st0012 — APPROVED** [review]

**Pull request merged**